colab_evaluation.py

```diff
@@ -0,0 +1,207 @@
+    """Create a tidy dataframe from ground truth data."""
+    up_ids = []
+    labels = []
+
+    for i in tqdm.tqdm(ground_truth.index, position=0):
+        up_id = ground_truth['sequence_name'][i]
+        for vocab_entry in ground_truth['true_label'][i]:
+            up_ids.append(up_id)
+            labels.append(vocab_entry)
+    return pd.DataFrame({"up_id": up_ids, "label": labels, "gt": True})
+
+
+def merge_predictions_and_ground_truth(predictions_df, ground_truth_df):
+    """Perform an inner join of predictions and ground truth, then set all empty values to False."""
+    combined = predictions_df.merge(ground_truth_df,
+                                    how="outer",
+                                    suffixes=("_pred", "_gt"),
+                                    left_on=["label", "up_id"],
+                                    right_on=["label", "up_id"])
+    combined = combined.fillna(False)
+    return combined
+
+
+def get_pr_curve_df(predictions_df,
+                    ground_truth_df,
+                    grouping=None,
+                    filtered=True):
+    """Given predictions and ground truth in tidy format, yield a precision recall curve.
+    
+    Args:
+        predictions_df: predictions in tidy format
+        ground_truth_df: ground truth in tidy format
+        grouping: optional dictionary mapping sequence names to categories
+        filtered: whether to remove almost redundant points on PR curve
+    """
+    combined = merge_predictions_and_ground_truth(predictions_df,
+                                                  ground_truth_df)
+    if grouping == None:
+        to_process = {'all': combined}.items()
+    else:
+        combined['group'] = combined['label'].map(grouping)
+        to_process = combined.groupby('group')
+
+    del combined
+    output_dfs = []
+    for group_name, group in tqdm.tqdm(to_process, position=0):
+        precisions, recalls, thresholds = sklearn.metrics.precision_recall_curve(
+            group['gt'], group['value'])
+        precisions = precisions[:-1]
+        recalls = recalls[:-1]
+        if filtered:
+            precisions, recalls, thresholds = filter_pr_curve(
+                precisions, recalls, thresholds)
+        output_dfs.append(
+            pd.DataFrame({
+                'group':
+                group_name,
+                'precision':
+                precisions,
+                'recall':
+                recalls,
+                'threshold':
+                thresholds,
+                'f1':
+                2 * precisions * recalls / (precisions + recalls)
+            }))
+    return pd.concat(output_dfs)
+
+
+def filter_pr_curve(precisions, recalls, thresholds, resolution=1e-3):
+    """Filters out imperceptible shifts in a PR curve."""
+    last_precision = None
+    last_recall = None
+    new_precisions = []
+    new_recalls = []
+    new_thresholds = []
+    for i in range(len(precisions)):

```

Comment:
Thanks - for next time. I'm not sure this is super straightforward to vectorise across the long dimension because I think you have to cache the last saved value (but maybe there is a neat way of doing it). Agree would be good to vectorise on the recall/precision direction.